ui/app/src/services/content.ts

```diff
@@ -562,6 +646,69 @@ export function insertComponent(
   );
 }
 
+export function insertInstance(
+  site: string,
+  modelId: string,
+  fieldId: string,
+  targetIndex: number,
+  instance: ContentInstance,
+  parentModelId: string = null,
+): Observable<any> {
+  return performMutation(
+    site,
+    modelId,
+    parentModelId,
+    doc => {
+      const qs = {
+        site,
+        path: modelId,
+        unlock: 'true',
+        fileName: getInnerHtml(doc.querySelector('file-name'))
+      };
+
+      const id = instance.craftercms.id;
+      const pathBase = `/site/components/${instance.craftercms.contentType.replace('/component/', '')}s/`.replace(/\/{1,}$/m, '');
+      const path = `${pathBase}/${id}.xml`;
+
+      const newItem = doc.createElement('item');
+
+      createElements(doc, newItem, {
+        '@attributes': {
+          // TODO: Hardcoded value. Fix.
+          datasource: 'sharedFeatures'
+        },
+        key: path,
+        value: instance.craftercms.label,
+        ...{
+          include: path,
+          disableFlattening: 'false'
+        }
+      });
+
+      let fieldNode = doc.querySelector(`:scope > ${fieldId}`);
+
+      if (nou(fieldNode)) {
+        fieldNode = doc.createElement(fieldId);
+        fieldNode.setAttribute('item-list', 'true');
+        doc.documentElement.appendChild(fieldNode);
+      }
+
+      const itemList = fieldNode.querySelectorAll(`:scope > item`);
+
+      if (itemList.length === targetIndex) {
+        fieldNode.appendChild(newItem);
+      } else {
+        $(newItem).insertBefore(itemList[targetIndex]);
+      }
+
+      return post(
+        writeContentUrl(qs),
+        serialize(doc)
+      );
```

Comment:
The mutation function should only perform the necessary document mutations. Shouldn't do any sort of serializing, writing, etc.